site/src/xServices/workspaceBuild/workspaceBuildXService.ts

```diff
@@ -87,16 +109,32 @@ export const workspaceBuildMachine = createMachine(
       assignLogs: assign({
         logs: (_, event) => event.data,
       }),
-      assignGetBuildLogsError: assign({
-        getBuildLogsError: (_, event) => event.data,
-      }),
-      clearGetBuildLogsError: assign({
-        getBuildLogsError: (_) => undefined,
+      addLog: assign({
+        logs: (context, event) => {
+          const previousLogs = context.logs ?? []
+          return [...previousLogs, event.log]
+        },
       }),
     },
     services: {
       getWorkspaceBuild: (ctx) => API.getWorkspaceBuild(ctx.buildId),
-      getWorkspaceBuildLogs: (ctx) => API.getWorkspaceBuildLogs(ctx.buildId),
+      getLogs: async (ctx) => API.getWorkspaceBuildLogs(ctx.buildId),
+      streamWorkspaceBuildLogs: (ctx) => async (callback) => {
+        const reader = await API.streamWorkspaceBuildLogs(ctx.buildId)
+
+        // Watching for the stream
+        // eslint-disable-next-line no-constant-condition, @typescript-eslint/no-unnecessary-condition
+        while (true) {
```

Comment:
It is a good point. I think for now we can rely on done being always called, why:
- We are using a polyfill(https://canjs.com/doc/can-ndjson-stream.html)  because Firefox does not support some stream APIs
-  I took a look and I didn't find anything related to "read failure" or something related
- If an error is thrown, it will stop the execution anyway
- This error, in particular, looks tricky to handle so I would expect someone having issues with that to think of a proper solution
- In summary, I don't know how to do that and it looks complex to handle, so I would not invest too much time on that right now, but if you have an idea how to solve that, I would love to learn and implement it for sure.

Reply:
Ha! Thanks for the breakdown - what you said makes sense and I think you're right to hold off. If this becomes an issue in the future, we can look into [calling `cancel` on the ReadableStream object](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/cancel).

Reply:
Ahh, nice!